query-graphs-js/src/querygraph.ts

```diff
@@ -163,27 +161,13 @@ export class Edge {
     return this.transition.kind === 'FieldCollection' && this.transition.definition.name === name;
   }
 
-  matchesSupergraphTransition(supergraph: Schema, otherTransition: Transition): boolean {
+  matchesSupergraphTransition(otherTransition: Transition): boolean {
+    assert(otherTransition.collectOperationElements, "Supergraphs shouldn't have transition that don't collect elements");
     const transition = this.transition;
     switch (transition.kind) {
-      case 'FieldCollection':
```

Comment:
What this method was previously trying to do is enforce the "intersection" semantic for input types, but to do that during composition validation. It is, in a way, the part that was broken and is fixed in this patch. 

The idea of this code was to merge inputs with a union (so, say, keeping any input type field defined in any subgraph), but then when doing composition validation for a field, to ignore any definition of that field that don't accept the "full" input object as merged in the supergraph API. Which really ends up requiring (in 99.99% of the case) that all the inputs are the same in all subgraphs, but is, in hindsight, an overly complex and confusing way to do this. Which is why this patch removes this and instead deal with input types during merging, which is easier to do.

Now, you can ignore this code since it's removed, but if you're wondering "I can admit this wasn't idea, but why wasn't this working?" then the answer is that there is a typo in the `isAccessible` method of `structuralSubtyping.ts` (a missing negation), which essentially made all this code a no-op. What happens if we fix that typo? Well, I tried and run into some assertion errors and I'm sure that assertion could been fixed easily enough but I didn't looked more closely because even if this code ended up working as I intended it to 6+ months ago when I wrote it, it would end up in pretty confusing composition validation error messages and we would have to work hard to make those errors understandable. That's one of the reason why dealing with this during merging is likely way better for input: it's much much easier to provide good error messages.